hw/drivers/debounce/include/debounce/debounce.h

```diff
@@ -0,0 +1,235 @@
+ */
+typedef struct debounce_s {
+    uint32_t pin     : 24;
+    uint32_t state   :  1;
+    uint32_t on_rise :  1;
+    uint32_t on_fall :  1;
+    uint32_t         :  0;
+    uint32_t ticks   : 16;
+    uint32_t count   :  8;
+    uint32_t accu    :  8;
+    void (*on_change)(struct debounce_s*);
+    void *arg;
+    struct hal_timer timer;
+} debounce_t;
+
+/*
+ * A callback can be invoked when the debounced pin's state changes. It
+ * can either be invoked when the pin rises, falls or on any state change.
+ * If either \c DEBOUNCE_CALLBACK_NEVER is specified or the callbak itself
+ * is set to \c NULL no callback is invoked.
+ */
+typedef enum debounce_callback_event_e {
+    DEBOUNCE_CALLBACK_NEVER      = 0, /* no callback invocation */
+    DEBOUNCE_CALLBACK_EVENT_RISE = 1, /* callbak when signal rises */
+    DEBOUNCE_CALLBACK_EVENT_FALL = 2, /* callbak when signal falls */
+    DEBOUNCE_CALLBACK_EVENT_ANY  = 3  /* callbak when signal changes */
+} debounce_callback_event_t;
+
+/*
+ * Prototype for the callback function. The callback has access to the pin,
+ * current state and the initially provided argument through the API.
+ * See also
+ *  - debounce_arg
+ *  - debounce_pin
+ *  - debounce_state
+ */
+typedef void (*debounce_callback_t)(debounce_t *);
+
+/**
+ * debounce init
+ *
+ * Initializes the specified pin to be debounced. Has to be called before
+ * any other debouncing function on the given \c debounce structure.
+ *
+ * @param d         Structure to manage debouncing
+ * @param pin       Pin number to set as input
+ * @param pull      Pull type, see hal_gpio.h
+ * @param timer     The HW timer number to be used by the debouncer.
+ *                  The timer has to be configured and setup properly by the
+ *                  application before this call.
+ *
+ * @return int  0: no error; -1 otherwise.
+ */
+int debounce_init(debounce_t *d, int pin, hal_gpio_pull_t pull, int timer);
+
+/**
+ * debounce set params
+ *
+ * Can optionally be used to tune the debouncing parameters.
+ * Once there is a pin change detected debouncing the checks the pin's state
+ * periodically and requires it to be identical for a certain number of times.
+ *
+ * The critical time introduced by debouncing can be calculated by
+ *   critical_time = timer_tick_period * ticks * count
+ * This value describes the minim latency introduce by debouncing a pin. It
+ * also represents the minimum time a pin has to be asserted or not for the
+ * change to be propagated.
+ *
+ * The default values are defined by package values
+ *   DEBOUNCE_PARAM_TICKS 1
+ *   DEBOUNCE_PARAM_COUNT 10
+ *
+ *
+ * @param d         Structure to manage debouncing
+ * @param ticks     The # ticks used as the time interval for debouncing
+ * @param count     The # times the pin state has to be stable for
+ *                  debouncing to complete
+ *
+ * @return int  0: no error; -1 otherwise.
+ */
+int debounce_set_params(debounce_t *d, uint16_t ticks, uint8_t count);
```

Comment:
Maybe debounce_config()? To use the same term as other drivers.

Reply:
The reason I didn't use `_config` is because in all drivers and APIs I looked at the `_config` call is mandatory, whereas setting these parameters in code is only necessary if you have multiple pins to debounce in one application and they need different debouncing parameters.